myokit/_model_api.py

```diff
@@ -1659,6 +1659,276 @@ def has_warnings(self):
+        """
+        Imports a component from another model (the "source model") into this
+        one (the "target model").
+
+        If variables in the ``external_component`` refer to variables from
+        other components they will be mapped on to variables from this model
+        by trying the following strategies (in order):
+
+        1. Using the provided dict ``var_map``, which maps variables in the
+           source model to variables in this model.
+        2. By assuming variables with the same label/binding map onto each
+           other.
+        3. By assuming that variables with the same qualified names map onto
+           each other. This rule is only applied if ``allow_name_mapping`` is
+           set to ``True``.
+
+        Arguments:
+
+        ``external_component``
+            A :class:`myokit.Component` from another model.
+        ``new_name``
+            An optional new name for the imported component.
+        ``var_map``
+            An optional dict mapping variables in the source model to
+            variables from this model (with variables specified as objects or
+            as fully qualified names).
+        ``allow_name_mapping``
+            Set this to ``True`` to allow mapping based on fully qualified
+            names, e.g. a variable ``x.y`` in the source model will be mapped
+            to ab variable ``x.y`` in the target model.
+        ``convert_units``
+            Set this to ``True`` to convert the units of any mapped variables,
+            if required and possible.
+
+        If any variables cannot be mapped, a :class:`myokit.WellMappedError`
+        will be raised. A :class:`myokit.DuplicateNameError` will be raised if
+        the components name or ``new_name`` clashes with an existing component
+        name. If unit conversion is enabled and variables with incompattible
+        units are mapped onto each other, a
+        class:`myokit.IncompatibleUnitsError` will be raised.
+        """
+
+        # Check component is not from this model, is a component, etc.
+        if not isinstance(external_component, myokit.Component):
+            raise TypeError(
+                'Method import_component() expects a myokit.Component'
+            )
+        if external_component.has_ancestor(self):
+            raise myokit.IntegrityError(
+                '<' + external_component.name() +
+                '> already has this model as an ancestor'
+            )
+
+        # Check if new name is provided, or else check that name doesn't clash
+        if new_name is None:
+            new_name = external_component.name()
+        if self.has_component(new_name):
+            raise myokit.DuplicateName(
+                'This model already has component with the name ' + new_name
+            )
+
+        # Create mapping for vars, and get unit conversion factors if required.
+
+        ext_model_copy = external_component.model().clone()
```

Comment:
Why the clones? (This is quite expensive)

Reply:
The clone is because I need to change some things about the external model (converting units and altering the structure of the model so the equations reference the correct variables). I do realise this is quite expensive but that was the easiest way I could find to do it without changing the external model.

Reply:
I have now implemented the function in a way that does not require cloning the model.